codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -358,18 +357,23 @@ public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, lon
      */
     @Override
     public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode, ByteBuf debugData,
-            ChannelPromise promise) {
-        Http2Connection connection = connection();
-        if (connection.isGoAway()) {
+                                     ChannelPromise promise) {
+        try {
+            if (connection().goAwaySent() && connection().remote().lastKnownStream() < lastStreamId) {
+                throw connectionError(PROTOCOL_ERROR, "Last stream identifier must not increase between " +
+                                                      "sending multiple GOAWAY frames (was '%d', is '%d').",
+                                                      connection().remote().lastKnownStream(),
+                                                      lastStreamId);
+            }
+            connection().goAwaySent(lastStreamId);
```

Comment:
In summary I think we can synchronously set the `lastStreamId` (to prevent others from quickly trying to send invalid goaway frames), and then asynchronously:
1. ON SUCCESS: notify listeners, and close all streams.
2. ON FAILURE: re-set the `lastStreamId`, log a failure.